---
title: "LINUX+R"

date: "2024-10-12"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Linuxå…¥é—¨

## 01_LinuxåŸºç¡€

### ç™»å½•æœåŠ¡å™¨

```{bash}
ç”¨æˆ·åã€å¯†ç å’Œipåœ°å€ï¼Œç™»å½•æ–¹å¼ä¸ºï¼š`ssh   ç”¨æˆ·å@ipåœ°å€`ï¼Œå¦‚ï¼š
ssh  vip28@94.191.82.93
#å›è½¦ï¼Œç„¶åè¾“å…¥å¯†ç 
```

### æŸ¥çœ‹å¸®åŠ©æ–‡æ¡£

```{bash}
#man å‘½ä»¤ï¼Œhelp å‘½ä»¤ï¼Œæˆ–è€…æŸä¸ªå‘½ä»¤çš„  --help  å‚æ•°
man  ls		## ç”¨ man å‘½ä»¤æŸ¥çœ‹ ls å‘½ä»¤çš„å¸®åŠ©æ–‡æ¡£
help  ls	## ç”¨ help å‘½ä»¤æŸ¥çœ‹ ls å‘½ä»¤çš„å¸®åŠ©æ–‡æ¡£	
ls  --help	## ç”¨ --help å‚æ•°æŸ¥çœ‹ ls å‘½ä»¤çš„å¸®åŠ©æ–‡æ¡£
```

### å¸¸è§çš„ç¯å¢ƒå˜é‡ï¼š\$HOME ï¼ŒPS1 \$PATH

```{bash}
#`$HOME` è®°å½•äº†ç”¨æˆ·çš„å®¶ç›®å½•æ‰€åœ¨çš„è·¯å¾„

#`PS1` å‘½ä»¤è¡Œé…è‰²
$ echo  $HOME
/trainee2/vip28

$ echo  $PS1
\[\033]2;\h:\u \w\007\033[33;1m\]\u \033[35;1m\t\033[0m \[\033[36;1m\]\w\[\033[0m\]\n\[\e[32;1m\]$ \[\e[0m\]

$ echo  $PATH
/trainee2/vip28/miniconda3/condabin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin

```

#### ä¿®æ”¹å‘½ä»¤è¡Œé…è‰²

```{bash}
#æ„Ÿå…´è¶£çš„è‡ªè¡Œæœç´¢
echo  'export PS1="\[\033]2;\h:\u \w\007\033[33;1m\]\u \033[35;1m\t\033[0m \[\033[36;1m\]\w\[\033[0m\]\n\[\e[32;1m\]$ \[\e[0m\]" ' >> ~/.bashrc
source  ~/.bashrc
#`~/.bashrc`ï¼šç³»ç»Ÿé…ç½®æ–‡ä»¶ï¼ŒåŒ…å«ä¸“ç”¨äºä½ çš„ bash shell çš„bashä¿¡æ¯ã€è®¾ç½®ï¼Œæ¯æ¬¡ç™»å½•æˆ–æ‰“å¼€æ–°çš„ shell æ—¶ï¼Œè¯¥æ–‡ä»¶ä¼šè¢«è‡ªåŠ¨è¯»å–å’Œæ‰§è¡Œã€‚
```

#### PATH

```{bash}
#`$PATH`ï¼šè¾“å…¥å‘½ä»¤æ—¶Linuxä¼šå»æŸ¥æ‰¾PATHé‡Œé¢è®°å½•çš„è·¯å¾„ï¼Œå¦‚æœå‘½ä»¤å­˜åœ¨æŸä¸€ä¸ªè·¯å¾„ä¸­ï¼Œå°±å¯ä»¥æˆåŠŸè°ƒç”¨ã€‚

#`<PATH1>:<PATH2>:<PATH3>:------:<PATHN>`

#æ‰“ä¸ªæ¯”æ–¹ï¼ŒPATH æ˜¯ä¸€ä¸ªå·¥å…·ç®±ï¼Œæœ‰å¾ˆå¤šå±‚ï¼ˆå¯¹åº”å¾ˆå¤šä¸ªè·¯å¾„ï¼‰ï¼Œæ¯ä¸€å±‚æ”¾ç€å„å¼å„æ ·çš„å·¥å…·ï¼ˆå¯¹åº”å„ç§å‘½ä»¤ï¼‰ã€‚
$ echo $PATH
/trainee2/vip28/miniconda3/condabin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin

# å¯ä»¥æŠŠ : æ›¿æ¢æˆæ¢è¡Œç¬¦ \n 
$ echo $PATH | tr ':'  '\n'
/trainee2/vip28/miniconda3/condabin
/usr/local/sbin
/usr/local/bin
/usr/sbin
/usr/bin
/sbin
/bin
/usr/games
/usr/local/games
/snap/bin

# æ¯”å¦‚ ls å‘½ä»¤å­˜åœ¨
$ ls  
$ which ls 
/bin/ls

#### å¦‚ä½•ç®¡ç† PATH

#å¦‚ä½•ç®¡ç† `$PATH`ï¼šç†è§£ç¯å¢ƒå˜é‡ `$PATH`  æ˜¯éå¸¸é‡è¦çš„ï¼Œå¯¹åç»­çš„ç¯å¢ƒå’Œè½¯ä»¶ç®¡ç†éƒ½éå¸¸é‡è¦ã€‚

#æ¨èæ–¹æ³•ï¼šåœ¨è‡ªå·±å®¶ç›®å½•ä¸‹åˆ›å»ºä¸€ä¸ª `~/bin/` æ–‡ä»¶å¤¹å¹¶å°†å…¶æ·»åŠ åˆ°ç¯å¢ƒå˜é‡ï¼Œåç»­å®‰è£…è½¯ä»¶ï¼Œå°±å°†è½¯ä»¶çš„å¯æ‰§è¡Œæ–‡ä»¶æ‹·è´æˆ–è½¯é“¾æ¥ï¼ˆç»å¯¹è·¯å¾„ï¼‰åˆ°è¿™ä¸ª bin æ–‡ä»¶å¤¹ï¼š
mkdir  ~/bin 
echo  'export "PATH=~/bin:$PATH" ' >> ~/.bashrc 
source  ~/.bashrc
```

### Linuxå¸¸ç”¨å‘½ä»¤

#### ls å‘½ä»¤

```{bash}
#åˆ—å‡ºç›®å½•æ–‡ä»¶æƒ…å†µ
ls				## åˆ—å‡ºå½“å‰ç›®å½•çš„æ–‡ä»¶
ls  ./			## åŒä¸Šï¼Œâ€˜.â€™å·ä»£è¡¨å½“å‰ç›®å½•
ls  ./*txt		## åˆ—å‡ºå½“å‰ç›®å½•ä¸‹ä»¥ txt ç»“å°¾çš„æ–‡ä»¶
ls  ../ 		## åˆ—å‡ºä¸Šå±‚ç›®å½•çš„æ–‡ä»¶
ls  -a			## åˆ—å‡ºå½“å‰ç›®å½•ä¸‹çš„æ‰€æœ‰æ–‡ä»¶ï¼ŒåŒ…æ‹¬éšè—æ–‡ä»¶
ls Â -l			## åˆ—å‡ºå½“å‰ç›®å½•ä¸‹æ–‡ä»¶çš„è¯¦ç»†ä¿¡æ¯
ll				## ls  -la çš„ç®€å†™
ls Â -lh 		## åŠ ä¸Š -h å‚æ•°ï¼Œä»¥ Kã€Mã€G çš„å½¢å¼æ˜¾ç¤ºæ–‡ä»¶å¤§å°
ls Â -lh  /		## åˆ—å‡ºæ ¹ç›®å½•ä¸‹æ–‡ä»¶çš„è¯¦ç»†ä¿¡æ¯
```

#### cd å‘½ä»¤

```{bash}
cd  .. Â  Â  Â  ## åˆ‡æ¢åˆ°ä¸Šå±‚ç›®å½•ï¼Œç›¸å¯¹è·¯å¾„
cd  / Â  Â  Â  Â ## åˆ‡æ¢åˆ°æ ¹ç›®å½•
cd  /teach/  ## åˆ‡æ¢åˆ°æ ¹ç›®å½•ä¸‹çš„teachï¼Œç»å¯¹è·¯å¾„
cd Â - Â  Â  Â   ## è¿”å›ä¸Šä¸€æ¬¡çš„å·¥ä½œç›®å½•
cd  ~ Â  Â  Â   ## å›åˆ°ç”¨æˆ·å®¶ç›®å½•
cd Â  Â  Â  Â  Â  ## åŒä¸Šï¼Œå›åˆ°ç”¨æˆ·å®¶ç›®å½•
```

#### mkdir

```{bash}
# åˆ›å»ºç›®å½•
mkdir dir0
ls
mkdir dir0/sub1/sub2
ls
ls dir0
mkdir -p dir0/sub1/sub2
ls dir0
ls dir0/sub1/
mkdir -p  test{1..3}/test{1..3}
tree
```

#### touch

```{bash}
ls
touch  file.txt  new.txt
ls
touch  file{1..5}
ls
```

#### rm

```         
rm  -i  file.txt
ls  file*
rm  file*
rm  -rf  test1
```

#### mv

```{bash}
mv  file1   Data/file2
```

#### cp

```{bash}
cp   readme.txt   Data/
mkdir  dir0
cp  -r  dir0  Data/

```

#### ln

```{bash}
ln -s /teach/software/Miniconda3-latest-Linux-x86_64.sh  ./

```

#### tar

```{bash}
## è§£å‹
tar  -zxvf  Data.tar.gz
## å‹ç¼©
tar  -zcvf  Data.tar.gz    Data  ...

```

#### cat

```{bash}
cat  readme.txt
cat  -n  readme.txt
## å†™å…¥æ–‡ä»¶
cat >file
Welcome to Biotrainee() !
^C			## è¿™é‡Œæ˜¯æŒ‰Crtl  C
## æŸ¥çœ‹
cat file
Welcome to Biotrainee() !

```

#### headã€tail

```{bash}
head  -n  20  Data/example.fq
## æŸ¥çœ‹ .bashrc çš„æœ€å 10 è¡Œ
tail  ~/.bashrc
## æŸ¥çœ‹ç¬¬20è¡Œ
head  -n  20  Data/example.fq | tail -1

```

#### less

æŒ‰ q é€€å‡º

```{bash}
less  Data/example.fq
less -S Data/example.fq
less -N Data/example.fq
zless -N Data/reads.1.fq.gz

```

#### wc

```{bash}
cat -n readme.txt  # æ˜¾ç¤ºreadme.txtæ–‡ä»¶å†…å®¹å¹¶åŠ ä¸Šè¡Œå·
cat readme.txt | wc  # è¾“å‡ºreadme.txtæ–‡ä»¶çš„å•è¯æ•°ã€è¡Œæ•°å’Œå­—èŠ‚æ•°
wc -l readme.txt  # åªè¾“å‡ºreadme.txtæ–‡ä»¶çš„è¡Œæ•°
```

#### cut

```{bash}
less -S Data/example.gtf | cut -f 1,3-5  # æ˜¾ç¤ºexample.gtfæ–‡ä»¶çš„ç¬¬1åˆ—å’Œç¬¬3åˆ°5åˆ—
less -S Data/example.gtf | cut -d 'h' -f 1  # ä½¿ç”¨'h'ä½œä¸ºåˆ†éš”ç¬¦ï¼Œæ˜¾ç¤ºexample.gtfæ–‡ä»¶çš„ç¬¬ä¸€åˆ—
```

#### sort

```{bash}
less -S Data/example.gtf | sort -k 4 | less -S  # æŒ‰ç¬¬4åˆ—æ’åºexample.gtfæ–‡ä»¶å†…å®¹åæ˜¾ç¤º
less -S Data/example.gtf | sort -n -k 4 | less -S  # æŒ‰ç¬¬4åˆ—æ•°å€¼æ’åºexample.gtfæ–‡ä»¶å†…å®¹åæ˜¾ç¤º

```

#### uniq

```{bash}
less -S Data/example.gtf | cut -f 3 | sort | uniq -c  # ç»Ÿè®¡example.gtfæ–‡ä»¶ç¬¬3åˆ—çš„å”¯ä¸€å€¼åŠå…¶å‡ºç°æ¬¡æ•°

```

#### paste

```{bash}
less -S Data/example.fq | paste - - - | less -S  # å°†example.fqæ–‡ä»¶çš„æ¯ä¸€è¡Œä¸ä¸‰ä¸ªç©ºè¾“å…¥åˆå¹¶æˆä¸€è¡Œåæ˜¾ç¤º
paste file1 file2  # å°†file1å’Œfile2çš„æ¯ä¸€è¡Œåˆå¹¶æˆä¸€è¡Œ
```

#### tr

```{bash}

cat readme.txt | tr 'e' 'E'  # å°†readme.txtæ–‡ä»¶ä¸­çš„'e'æ›¿æ¢ä¸º'E'
cat readme.txt | tr '\n' '\t'  # å°†readme.txtæ–‡ä»¶ä¸­çš„æ¢è¡Œç¬¦æ›¿æ¢ä¸ºåˆ¶è¡¨ç¬¦
cat readme.txt | tr -d 'e'  # åˆ é™¤readme.txtæ–‡ä»¶ä¸­çš„æ‰€æœ‰'e'
```

#### grep

```{bash}
#å¤šçœ‹çœ‹ç”Ÿä¿¡æŠ€èƒ½ä¹¦çš„ä¸‰é©¾é©¬è½¦
# grepï¼šä¸€ç§å¼ºå¤§çš„æ–‡æœ¬æœç´¢å·¥å…·ï¼Œå®ƒèƒ½ä½¿ç”¨æ­£åˆ™è¡¨è¾¾å¼åŒ¹é…æ¨¡å¼æœ
# ç´¢æ–‡æœ¬ï¼Œå¹¶æŠŠåŒ¹é…çš„è¡Œæ‰“å°å‡ºæ¥
# æ ¼å¼ï¼šgrep [options] pattern file
# å¸¸è§å‚æ•°ï¼š
# -wï¼šword ç²¾ç¡®æŸ¥æ‰¾æŸä¸ªå…³é”®è¯ pattern
# -cï¼šç»Ÿè®¡åŒ¹é…æˆåŠŸçš„è¡Œçš„æ•°é‡
# -vï¼šåå‘é€‰æ‹©ï¼Œå³è¾“å‡ºæ²¡æœ‰æ²¡æœ‰åŒ¹é…çš„è¡Œ
# -nï¼šæ˜¾ç¤ºåŒ¹é…æˆåŠŸçš„è¡Œæ‰€åœ¨çš„è¡Œå·
# -rï¼šä»ç›®å½•ä¸­æŸ¥æ‰¾pattern
# -eï¼šæŒ‡å®šå¤šä¸ªåŒ¹é…æ¨¡å¼
# -fï¼šä»æŒ‡å®šæ–‡ä»¶ä¸­è¯»å–è¦åŒ¹é…çš„ pattern
# -iï¼šå¿½ç•¥å¤§å°å†™

grep Biotrainee -r ./  # åœ¨å½“å‰ç›®å½•ä¸‹é€’å½’æœç´¢åŒ…å«'Biotrainee'çš„æ‰€æœ‰æ–‡ä»¶
less -S Data/example.fq | grep 'gene'  # åœ¨example.fqæ–‡ä»¶ä¸­æœç´¢åŒ…å«'gene'çš„è¡Œ
less -S Data/example.fq | grep -w 'gene'  # åœ¨example.fqæ–‡ä»¶ä¸­æœç´¢å®Œå…¨åŒ¹é…'gene'çš„è¡Œ
less -S Data/example.fq | grep -v -w 'gene'  # åœ¨example.fqæ–‡ä»¶ä¸­æœç´¢ä¸åŒ…å«'gene'çš„è¡Œ

```

#### æ­£åˆ™è¡¨è¾¾å¼

```{bash}
# ^ è¡Œé¦–
# $ è¡Œå°¾
# . æ¢è¡Œç¬¦ä¹‹å¤–çš„ä»»æ„å•ä¸ªå­—ç¬¦
# ? åŒ¹é…ä¹‹å‰é¡¹0æ¬¡æˆ–è€…ä¸€æ¬¡
# + åŒ¹é…1æ¬¡æˆ–è€…å¤šæ¬¡
# * åŒ¹é…0æ¬¡æˆ–è€…å¤šæ¬¡
# {n} åŒ¹é…næ¬¡
# {n,} åŒ¹é…è‡³å°‘næ¬¡
# {m,n} è‡³å°‘m,æœ€å¤šn
# [] åŒ¹é…ä»»æ„ä¸€ä¸ª
# [^] æ’é™¤å­—ç¬¦
# | æˆ–è€…
cat readme.txt  | grep '^T'  # åŒ¹é…ä»¥'T'å¼€å¤´çš„è¡Œ
cat readme.txt  | grep ')$'  # åŒ¹é…ä»¥')'ç»“å°¾çš„è¡Œ
cat readme.txt  | grep 'f.ee'  # åŒ¹é…å½¢å¦‚'fee', 'f1ee', 'f2ee'ç­‰çš„è¡Œï¼Œ`.` è¡¨ç¤ºä»»æ„å•ä¸ªå­—ç¬¦ã€‚
cat readme.txt  | grep 'f\?ee'  # åŒ¹é…'fee'æˆ–'free'ï¼Œ `\?` è¡¨ç¤ºé›¶ä¸ªæˆ–ä¸€ä¸ªå‰é¢çš„å­—ç¬¦ã€‚
cat readme.txt  | grep 're\+'  # åŒ¹é…ä¸€ä¸ªæˆ–å¤šä¸ª'r'åè·Ÿ'e'ï¼Œ`\+` è¡¨ç¤ºä¸€ä¸ªæˆ–å¤šä¸ªå‰é¢çš„å­—ç¬¦ã€‚
cat readme.txt  | grep [bB]  # åŒ¹é…åŒ…å«'b'æˆ–'B'çš„è¡Œï¼Œ`[bB]` è¡¨ç¤ºæ‹¬å·å†…çš„ä»»ä½•ä¸€ä¸ªå­—ç¬¦ã€‚
```

#### sed

```{bash}
# åœ¨readme.txtçš„ç¬¬ä¸€è¡Œå‰æ’å…¥ä¸€è¡Œ "Welcome to Biotrainee()"
cat readme.txt | sed '1i Welcome to Biotrainee() '

# åœ¨readme.txtçš„ç¬¬ä¸€è¡Œåè¿½åŠ ä¸€è¡Œ "Welcome to Biotrainee()"
cat readme.txt | sed '1a Welcome to Biotrainee() '

# å°†readme.txtçš„ç¬¬ä¸€è¡Œæ›¿æ¢ä¸º "Welcome to Biotrainee()"
cat readme.txt | sed '1c Welcome to Biotrainee()'

# å°†readme.txtä¸­çš„æ‰€æœ‰ "is" æ›¿æ¢ä¸º "IS"ï¼Œs/is/IS/g è¡¨ç¤ºå…¨å±€æ›¿æ¢ï¼ˆg ä»£è¡¨å…¨å±€ï¼‰ã€‚
cat readme.txt | sed 's/is/IS/g'

# åˆ é™¤readme.txtä¸­çš„æ‰€æœ‰ç©ºè¡Œï¼Œs/is/IS/g è¡¨ç¤ºå…¨å±€æ›¿æ¢ï¼ˆg ä»£è¡¨å…¨å±€ï¼‰ã€‚
cat readme.txt | sed '/^$/d'

# å°†readme.txtä¸­çš„æ‰€æœ‰ 'a' æ›¿æ¢ä¸º 'A'ï¼Œ'b' æ›¿æ¢ä¸º 'B'ï¼Œ'c' æ›¿æ¢ä¸º 'C'ï¼Œy/abc/ABC/ è¡¨ç¤ºå°† abc ä¸­çš„æ¯ä¸ªå­—ç¬¦æ›¿æ¢ä¸º ABC ä¸­å¯¹åº”çš„å­—ç¬¦ã€‚
cat readme.txt | sed 'y/abc/ABC/'
```

#### awk

```{bash}
# æ‰“å° example.gtf æ–‡ä»¶çš„ç¬¬9åˆ—
less -S Data/example.gtf | awk '{print $9}' | less -S

# æ‰“å° example.gtf æ–‡ä»¶çš„ç¬¬9åˆ—å’Œç¬¬10åˆ—
less -S Data/example.gtf | awk '{print $9, $10}' | less -S

# ä½¿ç”¨åˆ¶è¡¨ç¬¦ä½œä¸ºåˆ†éš”ç¬¦ï¼Œæ‰“å° example.gtf æ–‡ä»¶çš„ç¬¬9åˆ—
less -S Data/example.gtf | awk -F '\t' '{print $9}' | less -S

# å¦‚æœç¬¬3åˆ—æ˜¯ "gene"ï¼Œåˆ™æ‰“å°æ•´è¡Œ
less -S Data/example.gtf | awk '{if($3 == "gene") print $0}' | less -S

# å¦‚æœç¬¬3åˆ—æ˜¯ "gene"ï¼Œåˆ™æ‰“å°æ•´è¡Œï¼›å¦åˆ™æ‰“å° "$3 is not gene"
less -S Data/example.gtf | awk '{if($3 == "gene") {print $0} else {print $3 " is not gene "}}' | less -S

# å¦‚æœè¡Œä¸­åŒ…å« "gene"ï¼Œåˆ™æ‰“å°æ•´è¡Œ
less -S Data/example.gtf | awk '/gene/{print $0}' | less -S

# åœ¨å¼€å§‹æ—¶æ‰“å° "find UTR feature"ï¼Œå¦‚æœè¡Œä¸­åŒ…å« "UTR" åˆ™æ‰“å°æ•´è¡Œï¼Œåœ¨ç»“æŸæ—¶æ‰“å° "end"ï¼ŒBEGIN{print "find UTR feature"} åœ¨å¼€å§‹æ—¶æ‰“å° "find UTR feature"ã€‚/UTR/{print $0} å¦‚æœè¡Œä¸­åŒ…å« "UTR"ï¼Œåˆ™æ‰“å°æ•´è¡Œã€‚END{print "end"} åœ¨ç»“æŸæ—¶æ‰“å° "end"ã€‚
less -S Data/example.gtf | awk 'BEGIN{print "find UTR feature"} /UTR/{print $0} END{print "end"}'

# ä½¿ç”¨åˆ¶è¡¨ç¬¦ä½œä¸ºåˆ†éš”ç¬¦ï¼Œæ‰“å° example.gtf æ–‡ä»¶çš„ç¬¬9åˆ—ï¼ŒBEGIN{FS="\t"} è®¾ç½®åˆ¶è¡¨ç¬¦ä½œä¸ºåˆ†éš”ç¬¦ã€‚
less -S Data/example.gtf | awk 'BEGIN{FS="\t"} {print $9}' | less -S

# ä½¿ç”¨åˆ¶è¡¨ç¬¦ä½œä¸ºè¾“å…¥å’Œè¾“å‡ºåˆ†éš”ç¬¦ï¼Œå°†ç¬¬3åˆ—ä¸­çš„ "gene" æ›¿æ¢ä¸º "Gene"ï¼Œç„¶åæ‰“å°æ•´è¡Œï¼ŒBEGIN{FS="\t";OFS="\t"} è®¾ç½®åˆ¶è¡¨ç¬¦ä½œä¸ºè¾“å…¥å’Œè¾“å‡ºåˆ†éš”ç¬¦ã€‚ï¼Œ{gsub("gene", "Gene", $3); print $0} å°†ç¬¬3åˆ—ä¸­çš„ "gene" æ›¿æ¢ä¸º "Gene"ï¼Œç„¶åæ‰“å°æ•´è¡Œã€‚
less -S Data/example.gtf | awk 'BEGIN{FS="\t";OFS="\t"} {gsub("gene", "Gene", $3); print $0}' | less -S

```

### Linuxå¸¸ç”¨å‘½ä»¤åœºæ™¯ç»ƒä¹ 

#### åœºæ™¯ä¸€ï¼špwdã€cdã€ls ç»ƒä¹ 

> åœºæ™¯æ„å›¾ï¼š
>
> 1.  ç»ƒä¹ ä¸€æ¬¡è¿›å…¥å¤šå±‚æ–‡ä»¶å¤¹
>
> 2.  ä¹ æƒ¯ç”¨tabé”®è¡¥å…¨æ–‡ä»¶å’Œæ–‡ä»¶å¤¹
>
> 3.  ç†Ÿæ‚‰å‘½ä»¤çš„å„ç§å‚æ•°
>
> 4.  ç†Ÿæ‚‰æ–‡ä»¶æƒé™

1.  è¿›å…¥æ ¹ç›®å½•ä¸‹çš„varæ–‡ä»¶å¤¹ä¸‹çš„spoolæ–‡ä»¶å¤¹ï¼Œå¹¶æ‰“å°å‡ºå½“å‰ç›®å½•ä½ç½®, æŸ¥çœ‹å½“å‰æ–‡ä»¶å¤¹å†…å®¹

```{bash}
cd /var/spool/
pwd
ls
```

2.  ä¸€æ­¥å›åˆ°å®¶ç›®å½•(æä¾›è‡³å°‘3ç§è§£æ³•)

```{bash}
cd ~
cd 
cd /trainee2/Mar25 # æ­¤å¤„æ›¿æ¢æˆè‡ªå·±çš„å®¶ç›®å½•è·¯å¾„
```

3.  ä¸€æ­¥è¿”å›åˆšæ‰çš„æ–‡ä»¶å¤¹ï¼ˆæç¤ºï¼šä¸â€œ-â€æœ‰å…³ï¼‰

```{bash}
cd -
```

4.  æŸ¥çœ‹å®¶ç›®å½•ä¸‹çš„æ‰€æœ‰æ–‡ä»¶åŠæ–‡ä»¶å¤¹çš„è¯¦ç»†ä¿¡æ¯ï¼Œå›ç­”ï¼š.bashrcæ–‡ä»¶çš„æƒé™æ˜¯ï¼Ÿï¼ˆè°å¯è¯»ã€è°å¯å†™ã€è°å¯æ‰§è¡Œï¼‰

```{bash}
ls -la # æˆ–è€…ll -a
.bashrcæ–‡ä»¶æ‰€å±è€…å¯è¯»å¯å†™ä¸å¯æ‰§è¡Œï¼Œæ‰€å±ç»„å’Œå…¶ä»–äººå‡åªå¯è¯»ä¸å¯å†™ä¸å¯æ‰§è¡Œ
```

5.  ä¸€æ­¥è¿›å…¥å®¶ç›®å½•çš„ä¸Šå±‚ç›®å½•ä¸‹çš„ä½ çš„ç¼–å·Â±1çš„ç”¨æˆ·çš„ç›®å½•(å¦‚æœä¸æ»¡è¶³æ¡ä»¶ï¼Œåˆ™éšæœºè¿›å…¥ä¸¤ä¸ªå³å¯)

```{bash}
cd ../Mar24/
cd ../Mar26/
# æ­¤å¤„ä»¥Mar25ä¸ºä¾‹, åšé¢˜æ—¶æ›¿æ¢æˆä½ è‡ªå·±çš„ç”¨æˆ·åå³å¯ã€‚
```

#### åœºæ™¯äºŒï¼šmkdirã€touchã€treeç»ƒä¹ 

> å‡ºé¢˜æ„å›¾ï¼š
>
> 1.  å­¦ä¼šæ‰¹é‡åˆ›å»ºæ–‡ä»¶å’Œæ–‡ä»¶å¤¹
> 2.  ç†Ÿæ‚‰mkdirçš„é€‰é¡¹
> 3.  ç†Ÿæ‚‰treeå‘½ä»¤

1.  åœ¨å½“å‰ç›®å½•åˆ›å»ºå½¢å¦‚ 1/2/3/4/5/6/7/8/9 æ ¼å¼çš„æ–‡ä»¶å¤¹ç³»åˆ—

```{bash}
mkdir -p 1/2/3/4/5/6/7/8/9
```

2.  åœ¨ä¸ä½¿ç”¨cdå‘½ä»¤çš„å‰æä¸‹ï¼Œåœ¨ä¸Šä¸€é¢˜åˆ›å»ºçš„1/2/3/4/5/6/7/8/9ä¸‹ åˆ›å»ºæ–‡æœ¬æ–‡ä»¶findMe.txt

```{bash}
touch 1/2/3/4/5/6/7/8/9/findMe.txt
# æ­¤å¤„è®°å¾—ç”¨tabè¡¥é½å™¢
```

3.  ç”¨ä¸€æ¡å‘½ä»¤æ‰¹é‡åˆ›å»ºtestDir1\~10è¿™åä¸ªæ–‡ä»¶å¤¹

```{bash}
mkdir testDir{1..10}
```

4.  ç”¨ä¸€æ¡å‘½ä»¤åœ¨æ¯ä¸ªtestDir1\~10æ–‡ä»¶å¤¹ä¸­åˆ›å»ºä¸€ä¸ªmyFileæ–‡ä»¶

```{bash}
touch ./testDir{1..10}/myFile
```

5.  å°†è‡ªå·±çš„å®¶ç›®å½•æ–‡ä»¶å¤¹ä»¥æ ‘çš„ç»“æ„å±•ç¤ºå‡ºæ¥

```{bash}
tree ~
# æˆ–è€…å…ˆcdï¼Œåtree . ä¹Ÿå¯ã€‚ ä½†æ˜¯ä¸å¤Ÿç®€æ´
```

#### åœºæ™¯ä¸‰ï¼šmvã€cpå’Œrm

> å‡ºé¢˜æ„å›¾ï¼š
>
> 1.  ç†Ÿæ‚‰å¹¶åˆ†è¾¨mvçš„ç§»åŠ¨å’Œé‡å‘½ååŠŸèƒ½
> 2.  ç†Ÿæ‚‰cpå‘½ä»¤
> 3.  ç†Ÿæ‚‰rmå‘½ä»¤çš„äº¤äº’å¼å’Œé€’å½’åˆ é™¤

å‡†å¤‡å·¥ä½œï¼šåˆ›å»ºä¸¤ä¸ªæ–‡ä»¶file1å’Œfile2ï¼Œåˆ›å»ºä¸¤ä¸ªæ–‡ä»¶å¤¹myDir1å’ŒmyDir2ã€‚

1.  æŠŠfile1é‡å‘½åæˆfile3

```{bash}
mv file1 file3
```

2.  æŠŠfile2é‡å‘½åæˆfile3

```{bash}
mv file2 file3
```

3.  æŠŠfile3ç§»åŠ¨è¿›myDir1

```{bash}
mv file3 myDir1
```

4.  æŠŠmyDir1ç§»åŠ¨è¿›myDir2

```{bash}
mv myDir1 myDir2
```

5.  æŠŠmyDir2é‡å‘½åæˆmyDir3

```{bash}
mv myDir2 myDir3
```

6.  ç”¨ä¸€è¡Œå‘½ä»¤å°†1/2/3/4/5/6/7/8/9ä¸‹çš„findMe.txtæ–‡ä»¶å¤åˆ¶åˆ°å½“å‰æ–‡ä»¶å¤¹å¹¶å‘½åæˆfindMe

```{bash}
cp 1/2/3/4/5/6/7/8/9/findMe.txt ./fineMe
```

7.  æŠŠåœºæ™¯äºŒä¸­åˆ›å»ºçš„1/2/3/4/5/6/7/8/9å’ŒfindMe.txtæ–‡ä»¶ç”¨äº¤äº’å¼çš„æ–¹å¼åˆ é™¤

```{bash}
rm -ri 1
```

#### åœºæ™¯å››ï¼šln ç»ƒä¹ 

> å‡ºé¢˜æ„å›¾ï¼š
>
> å­¦ä¼šä½¿ç”¨ln -såˆ›å»ºè½¯è¿æ¥

å°†/home/t_linux/Miniconda3-latest-Linux-x86_64.shæ–‡ä»¶é“¾æ¥è‡³è‡ªå·±çš„å®¶ç›®å½•

```{bash}
cd ~
ln -s /home/t_linux/Miniconda3-latest-Linux-x86_64.sh .
```

#### åœºæ™¯äº”ï¼štarç»ƒä¹ 

> å­¦ä¼šç”¨tarè¿›è¡Œå‹ç¼©å’Œè§£å‹

1.  ç”¨tarå°†å®¶ç›®å½•ä¸‹çš„readme.txtæ–‡ä»¶å’Œè½¯è¿æ¥è¿‡æ¥çš„Miniconda3-latest-Linux-x86_64.shæ–‡ä»¶åˆ›å»ºå‹ç¼©æˆtest.tar.gzæ–‡ä»¶

```{bash}
tar -zcvf test.tar.gz readme.txt Miniconda3-latest-Linux-x86_64.sh
```

2.  åˆ›å»ºä¸€ä¸ªtestæ–‡ä»¶å¤¹ï¼ˆè‹¥å·²å­˜åœ¨åˆ™ä¸ç”¨åˆ›å»ºï¼‰ï¼Œå°†test.tar.gzæ–‡ä»¶ç§»åŠ¨åˆ°testä¸­å¹¶è§£å‹å¼€ã€‚

```{bash}
mkdir test
mv test.tar.gz
tar -zxvf test.tar.gz
```

## 02Linuxä¹‹å®‰è£…conda

### 1ä¸‹è½½

```{bash}
cd ~
wget -c https://repo.anaconda.com/archive/Anaconda3-2024.06-1-Linux-x86_64.sh
#åœ¨condaæ–‡ä»¶çš„ç›®å½•ä¸‹è¾“å…¥å‘½ä»¤å®‰è£…ï¼Œä¸€è·¯å›è½¦ï¼Œç›´åˆ°ä»–è¦æ±‚è¾“å…¥yes
bash Anaconda3-2024.06-1-Linux-x86_64.sh
#åœ¨æœ«å°¾æ·»åŠ ç¯å¢ƒå˜é‡
vim ~/.bashrc
export PATH=~/anaconda3/bin:$PATH
#åˆ·æ–°ç¯å¢ƒå˜é‡
source ~/.bashrc
#æœ€åconda -Vè¦æ˜¯æ­£å¸¸å°±å®‰è£…æˆåŠŸäº†
```

### 2é…ç½®é•œåƒæº

```{bash}
vim ~/.condarc
channels:
  - https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/free/
  - https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main/
  - https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/pytorch/
  - https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/msys2/
show_channel_urls: true
 
ssl_verify: true
allow_conda_downgrades: true


####
## é…ç½®é•œåƒ

# ä¸‹é¢å››è¡Œé…ç½®åŒ—äº¬å¤–å›½è¯­å¤§å­¦çš„condaçš„channelåœ°å€ï¼ˆé¦–é€‰ï¼‰
conda config --add channels https://mirrors.bfsu.edu.cn/anaconda/pkgs/main/ 
conda config --add channels https://mirrors.bfsu.edu.cn/anaconda/cloud/conda-forge/ 
conda config --add channels https://mirrors.bfsu.edu.cn/anaconda/cloud/bioconda/ 
conda config --set show_channel_urls yes 

# ä¸‹é¢è¿™å››è¡Œé…ç½®æ¸…åå¤§å­¦çš„condaçš„channelåœ°å€ï¼ˆé¦–é€‰åŒ—å¤–ï¼Œå¦‚æœä½“éªŒä¸å¥½å†æ¢æˆæ¸…åï¼‰
conda config --add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main/
conda config --add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge/
conda config --add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/bioconda/
conda config --set show_channel_urls yes

# å¦‚æœéœ€è¦å®˜æ–¹é¢‘é“ï¼Œå¯ä»¥æ·»åŠ ä¸‹é¢è¿™ä¸¤è¡Œé…ç½®å®˜ç½‘çš„channelåœ°å€ï¼ˆä¸æ¨èï¼‰
conda config --add channels conda-forge 
conda config --add channels bioconda

# åˆ é™¤defaultsé¢‘é“
sed -i '/defaults/d' ~/.condarc

## é…ç½®é•œåƒæˆåŠŸ
# æŸ¥çœ‹é…ç½®ç»“æœ
cat ~/.condarc
```

### 3åˆ›å»ºç¯å¢ƒï¼Œå¹¶å®‰è£…rstudio

```{bash}
conda create -n rstudio r=4.2.0 -y #æœ‰ç‰ˆæœ¬è¦æ±‚ï¼Œè¯•äº†åˆ«çš„ç‰ˆæœ¬ä¸æˆåŠŸ
conda activate rstudio
#å®‰è£…RStudioï¼š
conda install -c r rstudio --yes
#å®‰è£…RåŒ…
options(repos="http://mirrors.tuna.tsinghua.edu.cn/CRAN/")
options(BioC_mirror="http://mirrors.tuna.tsinghua.edu.cn/bioconductor/")
install.packages("devtools")
install.packages("BiocManager")
#è¿è¡Œ
rstudio
```

## 03Linuxä¹‹è½¯ä»¶å®‰è£…

### åˆ›å»ºå°ç¯å¢ƒ

```{bash}
# åˆ›å»ºåä¸ºrnaçš„è½¯ä»¶ç¯å¢ƒæ¥å®‰è£…è½¬å½•ç»„å­¦åˆ†æçš„ç”Ÿç‰©ä¿¡æ¯å­¦è½¯ä»¶
conda create -y -n  rna  python=3.7
# åˆ›å»ºå°ç¯å¢ƒæˆåŠŸï¼Œå¹¶æˆåŠŸå®‰è£…python3ç‰ˆæœ¬
# æ¯å»ºç«‹ä¸€ä¸ªå°ç¯å¢ƒï¼Œå®‰è£…ä¸€ä¸ªpython=3çš„è½¯ä»¶ä½œä¸ºä¾èµ–

# æŸ¥çœ‹å½“å‰condaç¯å¢ƒ
conda info -e
conda env list

# æ¯æ¬¡è¿è¡Œå‰ï¼Œæ¿€æ´»åˆ›å»ºçš„å°ç¯å¢ƒrna
conda activate rna

# é€€å‡ºå°ç¯å¢ƒ
conda deactivate
```

### åœ¨å°ç¯å¢ƒä¸­å®‰è£…ç”Ÿä¿¡è½¯ä»¶â€”â€”ä½¿ç”¨conndaå’Œmamba

```{bash}
# æ¿€æ´»ç¯å¢ƒ
#æ³¨ï¼šè½¯ä»¶éƒ½è¦å®‰è£…åœ¨å°ç¯å¢ƒä¸­ï¼Œä¸è¦å®‰è£…åœ¨ base 
### rnaç¯å¢ƒ
conda activate rna
# å®‰è£… fastqc è½¯ä»¶
conda  install  fastqc

# è°ƒå‡ºå¸®åŠ©æ–‡æ¡£
fastqc --help

# å¯ä»¥æŒ‡å®šè½¯ä»¶ç‰ˆæœ¬
conda install -y samtools=1.14 

# aspera 
conda install -y -c hcc aspera-cli
ascp --help

# å¯ä»¥ä¸€æ¬¡å®‰è£…å¤šä¸ªè½¯ä»¶
conda install -y python=3.7 libstdcxx-ng=9.1.0 trim-galore  hisat2  subread  multiqc  samtools=1.14  salmon=1.4.0 fastp fastqc
# mamba install -y python=3.7 libstdcxx-ng=9.1.0 trim-galore  hisat2  subread  multiqc  samtools=1.14  salmon=1.4.0 fastp fastqc

## ä¸æ˜¯é€šè¿‡è½¯ä»¶åæ¥è°ƒç”¨å¸®åŠ©æ–‡æ¡£ï¼Œè€Œæ˜¯è½¯ä»¶çš„å‘½ä»¤
# sra-tools
prefetch --help
fastq-dump --help
which prefetch

#  trim-galore
trim_galore --help

# hisat2
hisat2 --help

# subread
featureCounts --help

# multiqc
multiqc --help

# samtools
samtools --help

# salmon
salmon --help

# fastp
fastp --help

### R4ç¯å¢ƒ

# åˆ›å»ºR4ç¯å¢ƒ
conda create -y -n R4 python=3.8

# æ¿€æ´»R4ç¯å¢ƒ
conda activate R4

# (å¯é€‰æ­¥éª¤ï¼šåœ¨R4é‡Œå®‰è£…mamba)
# conda install mamba

# å®‰è£…Rè¯­è¨€æœ¬ä½“
conda install -y r-base=4.1.2
## æˆ–è€…ä½¿ç”¨mambaå®‰è£…ï¼š mamba install -y r-base=4.1.2

# å®‰è£…Rè¯­è¨€è½¯ä»¶åŒ…
conda install -y r-getopt r-tidyverse r-ggplot2=3.3.5 bioconductor-limma bioconductor-edger bioconductor-deseq2 bioconductor-clusterprofiler bioconductor-org.hs.eg.db=3.13.0
## æˆ–è€…ä½¿ç”¨mambaå®‰è£…ï¼šmamba install -y r-getopt r-tidyverse r-ggplot2=3.3.5 bioconductor-limma bioconductor-edger bioconductor-deseq2 bioconductor-clusterprofiler bioconductor-org.hs.eg.db=3.13.0
```

```{bash}
#Q: å¦‚ä½•éªŒè¯Rè¯­è¨€çš„åŒ…å®‰è£…æƒ…å†µï¼Ÿ
#A: è¿›å…¥Rè¯­è¨€ç¯å¢ƒä¸­ç”¨`library()` 
# è¾“å…¥Rè¿›å…¥Rè¯­è¨€çš„äº¤äº’
R
#åœ¨Rè¯­è¨€é‡ŒéªŒè¯å®‰è£…åŒ…ï¼š
library(getopt)
library(tidyverse)
library(ggplot2)
library(limma)
library(edgeR)
library(DESeq2)
library(clusterProfiler)
library(org.Hs.eg.db)
<!-- Q: å¦‚ä½•çŸ¥é“è¿™äº›åŒ…åˆ°åº•å«å•¥ï¼Ÿå“ªå„¿è¯¥å¤§å†™å“ªå„¿è¯¥å°å†™ï¼Ÿ -->

<!-- A: [Bioconductor - Home](https://bioconductor.org/) åœ¨Bioconductorçš„å®˜ç½‘æœç´¢å³å¯ã€‚ -->
```

### åœ¨å°ç¯å¢ƒä¸­å®‰è£…ç”Ÿä¿¡è½¯ä»¶â€”â€”ç›´æ¥å¯¼å…¥yamlé…ç½®æ–‡ä»¶ä»¥å®‰è£…è½¯ä»¶

```{bash}
conda env create -n rna -f rna.yaml
# å¦‚æœæœ‰mambaçš„è¯å¯ä»¥ç”¨mambaå®‰è£…
# mamba env create -n rna -f rna.yaml
conda env create -n R4 -f R4.yaml
```

### condaå…¶ä»–ç”¨æ³•æ›´æ–°è½¯ä»¶ï¼šconda update è½¯ä»¶åï¼Œå¸è½½è½¯ä»¶ï¼Œåˆ é™¤ç¯å¢ƒï¼Œå…‹éš†ç¯å¢ƒï¼ŒæŸ¥æ‰¾è½¯ä»¶

```{bash}
# æ›´æ–°è½¯ä»¶ï¼šconda  update  è½¯ä»¶å
# å¸è½½è½¯ä»¶ï¼šconda  removRe  è½¯ä»¶å
# åˆ é™¤ç¯å¢ƒï¼šconda  remove  -n   ç¯å¢ƒå
# å…‹éš†ç¯å¢ƒï¼šconda  create â€“n æ–°ç¯å¢ƒå â€“clone  æ—§ç¯å¢ƒå
# æŸ¥æ‰¾è½¯ä»¶ï¼šconda  search  è½¯ä»¶å
# æŸ¥æ‰¾è½¯ä»¶å¸¸ç”¨çš„é“¾æ¥ï¼š
# 
# - https://anaconda.org/search
# 
# - [https://bioconda.github.io](https://bioconda.github.io/)[/](https://bioconda.github.io/)

```

### æ‰‹åŠ¨å®‰è£…å®‰è£…è½¯ä»¶ä»¥åŠè½¯è¿æ¥

```{bash}
mkdir ~/biosoft 
cd ~/biosoft
# wget -c https://cloud.biohpc.swmed.edu/index.php/s/oTtGWbWjaxsQ2Ho/download -O hisat2-2.2.1-Linux_x86_64.zip
ln  -s  /teach/software/hisat2-2.2.1-Linux_x86_64.zip  ./
unzip hisat2-2.2.1-Linux_x86_64.zip
cd hisat2-2.2.1/
./hisat2 --help
# echo 'export PATH="${HOME}/biosoft/hisat2-2.2.1/:$PATH" ' >> ~/.bashrc 
ln  -s  ~/biosoft/hisat2-2.2.1/hisat2*   ~/bin/
#ä¹Ÿå¯ä»¥å°†å…¶ä»–ç¯å¢ƒä¸‹çš„Ræˆ–è€…å…¶ä»–è¯­è¨€æ‹·åˆ°~/bin/ï¼Œå¯ä»¥ä½¿å¾—åœ¨å®¶ç›®å½•ä¹Ÿå¯ä»¥è¿è¡Œ
ln -s ~/anaconda3/envs/rstudio/bin/R ~/bin/

```

## Linuxç³»ç»Ÿç¯å¢ƒ+Linuxç¼–ç¨‹

### æ–‡ä»¶ç³»ç»Ÿç»“æ„

```{bash}
/ 		è™šæ‹Ÿç›®å½•çš„æ ¹ç›®å½•ã€‚é€šå¸¸ä¸ä¼šåœ¨è¿™é‡Œå­˜å‚¨æ–‡ä»¶
/bin	äºŒè¿›åˆ¶ç›®å½•ï¼Œå­˜æ”¾è®¸å¤šç”¨æˆ·çº§çš„GNUå·¥å…·
/boot	å¯åŠ¨ç›®å½•ï¼Œå­˜æ”¾å¯åŠ¨æ–‡ä»¶
/dev	è®¾å¤‡ç›®å½•ï¼ŒLinuxåœ¨è¿™é‡Œåˆ›å»ºè®¾å¤‡èŠ‚ç‚¹
/etc	ç³»ç»Ÿé…ç½®æ–‡ä»¶ç›®å½•
/home	ä¸»ç›®å½•ï¼ŒLinuxåœ¨è¿™é‡Œåˆ›å»ºç”¨æˆ·ç›®å½•
/lib	åº“ç›®å½•ï¼Œå­˜æ”¾ç³»ç»Ÿå’Œåº”ç”¨ç¨‹åºçš„åº“æ–‡ä»¶
/media	åª’ä½“ç›®å½•ï¼Œå¯ç§»åŠ¨åª’ä½“è®¾å¤‡çš„å¸¸ç”¨æŒ‚è½½ç‚¹
/root	rootç”¨æˆ·çš„ä¸»ç›®å½•
/sbin	ç³»ç»ŸäºŒè¿›åˆ¶ç›®å½•ï¼Œå­˜æ”¾è®¸å¤šGNUç®¡ç†å‘˜çº§å·¥å…·
/run	è¿è¡Œç›®å½•ï¼Œå­˜æ”¾ç³»ç»Ÿè¿ä½œæ—¶çš„è¿è¡Œæ—¶æ•°æ®
/tmp	ä¸´æ—¶ç›®å½•ï¼Œå¯ä»¥åœ¨è¯¥ç›®å½•ä¸­åˆ›å»ºå’Œåˆ é™¤ä¸´æ—¶å·¥ä½œæ–‡ä»¶
/usr	ç”¨æˆ·äºŒè¿›åˆ¶ç›®å½•ï¼Œå¤§é‡ç”¨æˆ·çº§çš„GNUå·¥å…·å’Œæ•°æ®æ–‡ä»¶éƒ½å­˜å‚¨åœ¨è¿™é‡Œ
```

### æŸ¥çœ‹ç³»ç»Ÿèµ„æº

```{bash}

æŸ¥çœ‹CPUä¿¡æ¯ï¼šlscpu
æŸ¥çœ‹å†…å­˜ä¿¡æ¯ï¼šfree  -h
æŸ¥çœ‹ç¡¬ç›˜ä¿¡æ¯ï¼šdf  -h
æŸ¥çœ‹æ–‡ä»¶å¤§å°ï¼šdu  -h  -d  1
æŸ¥çœ‹æ–‡ä»¶å¤§å°ï¼šdu  -h  -d  1
æŸ¥çœ‹ç³»ç»Ÿè¿›ç¨‹ï¼štop æˆ–è€… ps -ef æˆ–è€… jobs

```

### å˜é‡

```{bash}

ç¯å¢ƒå˜é‡ã€çŠ¶æ€å˜é‡ã€ä½ç½®å‚æ•°å˜é‡ã€è‡ªå®šä¹‰å˜é‡ï¼Œè°ƒç”¨å˜é‡æ—¶ï¼Œè¦åœ¨å˜é‡å‰é¢åŠ ä¸€ä¸ª `$ `ç¬¦å·

ç¯å¢ƒå˜é‡ï¼šç”¨äºå­˜å‚¨æœ‰å…³shellä¼šè¯å’Œå·¥ä½œç¯å¢ƒçš„ç³»ç»Ÿå˜é‡

çŠ¶æ€å˜é‡ï¼šç”¨äºè®°å½•å‘½ä»¤çš„è¿è¡Œç»“æœ

ä½ç½®å‚æ•°å˜é‡ï¼šç”¨äºç”¨äºå‘å‘½ä»¤æˆ–ç¨‹åºè„šæœ¬ä¸­ä¼ é€’ä¿¡æ¯

è‡ªå®šä¹‰å˜é‡ï¼šç”±ç”¨æˆ·è‡ªè¡Œå®šä¹‰çš„å˜é‡ï¼Œå¯ç”¨äºç”¨æˆ·ç¼–å†™çš„è„šæœ¬ï¼Œå¤šä¸ªå‘½ä»¤é—´çš„å€¼ä¼ é€’ç­‰
```

### ç»“æ„åŒ–è¯­å¥

#### æ¡ä»¶è¯­å¥ if

```{bash}
#æœ‰å¤´æœ‰å°¾ï¼Œä¸€ä¸ª` if ` å°±è¦å¯¹åº”ä¸€ä¸ª ` fi ` ã€‚æœ‰ä¸‰ç§ç»“æ„ï¼š
# 1
if [ condition ]
then
	commands
fi

# 2
if [ condition ]
then
	commands
else
	commands
fi

# 3
if [ condition ]
then
	commands
else
	if [ condition ]
	then
		commands
	fi
fi
# (else if å¯ä»¥ç¼©å†™ä¸º elif )
if [ condition ]
then
	commands
elif [ condition ]
	then
		commands
	fi
fi

#ç¤ºä¾‹ï¼š
# æ•°å€¼åˆ¤æ–­
if [ 1 -eq 1 ]
then
  echo  "Welcome to Biotrainee() !"
else
  echo  "**************"
fi

# ç»“åˆçŠ¶æ€å‚æ•° $?
if [ $? -eq 0 ]
then
  touch  ok.txt
fi

# æ–‡ä»¶åˆ¤æ–­
if [ ! -f ok.txt ]
then
  touch  ok.txt
fi

```

#### for å¾ªç¯è¯­å¥

```{bash}
for i in  1 2 3 4 5 
do
  echo ${i} "Welcome to Biotrainee() !"
done

for i in {1..10}
do
  touch  file${i}
done

list="CDS exon gene start_codon stop_codon transcript UTR"
for i in ${list}
do
  echo  "This feature is ${i}"
done
```

#### while å¾ªç¯

```{bash}
ls file* | while  read  id;
do 
  mv ${id} ${id}.txt ; 
done


ls  file* > config
cat config | while  read id
do 
  mv  ${id}  ${id%.txt}
done
```

#### ç»“æ„åŒ–è¯­å¥ç»ƒä¹ é¢˜

```{bash}
id=example
fastqc ~/Data/${id}.fq
if [ $? -eq 0 ]
then
  echo "yes"
else 
  echo "no"
fi
```

```{bash}
touch  file{1..10}

ls file* | while read id
do 
  echo  "xxx"  > ${id}
done

```

### shell è„šæœ¬ä»¥åŠè„šæœ¬è¿è¡Œ

#### test.sh

```{bash}
$ vim  test.sh
#!/bin/bash
echo "Welcome to Biotrainee() !"

$ bash  test.sh

# 1æ ‡å‡†è¾“å‡º  å’Œ  2æ ‡å‡†è¯¯è¾“å‡ºï¼Œä½†æœ‰äº›è½¯ä»¶ä¸è§„èŒƒï¼Œæ‰€æœ‰è¾“å‡ºéƒ½åœ¨2
$ bash  test.sh  1>test.log  2>&1


# å¯æ‰§è¡Œæƒé™
$ ls -lh test.sh 

$ chmod  764 test.sh 

$ ls -lh test.sh 

# è·¯å¾„è°ƒç”¨å¯æ‰§è¡Œæ–‡ä»¶
./test.sh  

```

#### test2.sh

```{bash}
$ cat  test2.sh
#!/bin/bash
cat  $1

$ bash  test2.sh  readme.txt

```

#### test3.sh

```{bash}
$ cat  test3.sh
#!/bin/bash
echo  "Start"
sleep  100s
echo  "End"



```

#### shellè„šæœ¬åå°è¿è¡Œ

```{bash}
$ bash  test3.sh  

$ nohup  bash  test3.sh  &

$ nohup  bash  test3.sh   1>test3.log  2>&1  &

$ top

$ ps -ef | grep test3

```

### åœ¨ Linux ä¸­ä½¿ç”¨å…¶ä»–ç¼–ç¨‹è¯­è¨€

#### R è¯­è¨€è„šæœ¬

```{bash}
$ cat test.R 
#!/usr/bin/Rscript
a = 1:10
paste0("gene",a)

$ Rscript test.R 
 [1] "gene1"  "gene2"  "gene3"  "gene4"  "gene5"  
 [6] "gene6"  "gene7"  "gene8"  "gene9"  "gene10"

```

#### Python è„šæœ¬

```{bash}
$ cat test.py 
#!/usr/bin/python3

print("Hello World")

$ python3 test.py 
Hello World

```

### æ‰©å±•:è‡ªå·±å†™çš„å‘½ä»¤å®ç°è¡Œåˆ—è½¬æ¢ï¼ŒåŠŸèƒ½ç±»ä¼¼ R è¯­è¨€ä¸­çš„ `t( )` å‡½æ•°

```{bash}
$ cat  > row2col
awk 'BEGIN{FS="\t";OFS="\t"}{i=1;while(i <= NF){col[i]=col[i] $i "\t";i=i+1}} END {i=1;while(i<=NF){print i,col[i];i=i+1}}'

$ chmod 764  row2col

$ mv  row2col  ~/bin

$ head -n 2 ~/Data/example.gtf  |  row2col
```

# Rè¯­è¨€å…¥é—¨

## Rè¯­è¨€åŸºç¡€

### 1.å‘é‡

```{r}

# 1.å‘é‡ç”ŸæˆğŸŒŸ#####
#(1)ç”¨ c() ç»“åˆåˆ°ä¸€èµ·
c(2,5,6,2,9) 
c("a","f","md","b")
#(2)è¿ç»­çš„æ•°å­—ç”¨å†’å·â€œ:â€ 
1:5
#(3)æœ‰é‡å¤çš„ç”¨rep(),æœ‰è§„å¾‹çš„åºåˆ—ç”¨seq(),éšæœºæ•°ç”¨rnorm()
rep("x",times=3)  
seq(from=3,to=21,by=3)
rnorm(n=3)
#(4)é€šè¿‡ç»„åˆ,äº§ç”Ÿæ›´ä¸ºå¤æ‚çš„å‘é‡ã€‚
paste0(rep("x",times=3),1:3)

#####2.2å¯¹å•ä¸ªå‘é‡è¿›è¡Œçš„æ“ä½œ####
#(1)èµ‹å€¼ç»™ä¸€ä¸ªå˜é‡å
x = c(1,3,5,1) #éšæ„çš„å†™æ³•
x
x <- c(1,3,5,1) #è§„èŒƒçš„èµ‹å€¼ç¬¦å· Alt+å‡å·
x

#èµ‹å€¼+è¾“å‡ºä¸€èµ·å®ç°
x <- c(1,3,5,1);x
(x <- c(1,3,5,1))

#(2)ç®€å•æ•°å­¦è®¡ç®—
x+1
log(x)
sqrt(x)

#(3)æ ¹æ®æŸæ¡ä»¶è¿›è¡Œåˆ¤æ–­,ç”Ÿæˆé€»è¾‘å‹å‘é‡
x>3
x==3
#(4)åˆçº§ç»Ÿè®¡
max(x) #æœ€å¤§å€¼
min(x) #æœ€å°å€¼
mean(x) #å‡å€¼
median(x) #ä¸­ä½æ•°
var(x) #æ–¹å·®
sd(x) #æ ‡å‡†å·®
sum(x) #æ€»å’Œ

length(x) #é•¿åº¦
unique(x) #å»é‡å¤
duplicated(x) #å¯¹åº”å…ƒç´ æ˜¯å¦é‡å¤
table(x) #é‡å¤å€¼ç»Ÿè®¡
sort(x)
sort(x,decreasing = F)
sort(x,decreasing = T)
#####2.3.å¯¹ä¸¤ä¸ªå‘é‡è¿›è¡Œçš„æ“ä½œ#####
x = c(1,3,5,1)
y = c(3,2,5,6)
#(1)æ¯”è¾ƒè¿ç®—ï¼Œç”Ÿæˆç­‰é•¿çš„é€»è¾‘å‘é‡
x == y 
y == x
#(2)æ•°å­¦è®¡ç®—
x + y
#(3)è¿æ¥
paste(x,y,sep=",")

#pasteä¸paste0çš„åŒºåˆ«
paste(x,y)

paste0(x,y)
paste(x,y,sep = "")

paste(x,y,sep = ",")
#å½“ä¸¤ä¸ªå‘é‡é•¿åº¦ä¸ä¸€è‡´
x = c(1,3,5,6,2)
y = c(3,2,5)
x == y # Warning: é•¿çš„å¯¹è±¡é•¿åº¦ä¸æ˜¯çŸ­çš„å¯¹è±¡é•¿åº¦çš„æ•´å€æ•°[1] FALSE FALSE  TRUE FALSE  TRUE
#å¾ªç¯è¡¥é½

#åˆ©ç”¨å¾ªç¯è¡¥é½ç®€åŒ–ä»£ç 
paste0(rep("x",3),1:3)
paste0("x",1:3)

#(4)äº¤é›†ã€å¹¶é›†ã€å·®é›†
intersect(x,y)
union(x,y)
setdiff(x,y)
setdiff(y,x)

x %in% y #xçš„æ¯ä¸ªå…ƒç´ åœ¨yä¸­å­˜åœ¨å—
y %in% x #yçš„æ¯ä¸ªå…ƒç´ åœ¨xä¸­å­˜åœ¨å—

#####2.4.å‘é‡ç­›é€‰(å–å­é›†)--çœ‹ppt#####

x <- 8:12
#æ ¹æ®é€»è¾‘å€¼å–å­é›†
x[x==10]
x[x<12]
x[x %in% c(9,13)]
#æ ¹æ®ä½ç½®å–å­é›†
x[4]
x[2:4]
x[c(1,5)]
x[-4]
x[-(2:4)]

####2.5.ä¿®æ”¹å‘é‡ä¸­çš„æŸä¸ª/æŸäº›å…ƒç´ ï¼šå–å­é›†+èµ‹å€¼
x
#æ”¹ä¸€ä¸ªå…ƒç´ 
x[4] <- 40
x
#æ”¹å¤šä¸ªå…ƒç´ 
x[c(1,5)] <- c(80,20)
x

#### 2.6 ç®€å•å‘é‡ä½œå›¾
k1 = rnorm(12);k1
k2 = rep(c("a","b","c","d"),each = 3);k2
plot(k1)
boxplot(k1~k2) #è¯¾åè¯•ç€æœç´¢boxplotè¡¨è¾¾ä»€ä¹ˆæ„æ€


# 1.ç”Ÿæˆ1åˆ°15ä¹‹é—´æ‰€æœ‰å¶æ•°
seq(from = 1,to = 15,by = 2)
seq(from = 2,to = 15,by = 2)
# 2.ç”Ÿæˆå‘é‡ï¼Œå†…å®¹ä¸ºï¼š"student2"  "student4"  "student6"  "student8"  "student10" "student12" "student14" 
# æç¤ºï¼špaste0
paste0(rep("student",times = 7),seq(from = 2, to = 15,by = 2))
# 3.å°†ä¸¤ç§ä¸åŒç±»å‹çš„æ•°æ®ç”¨c()ç»„åˆåœ¨ä¸€èµ·ï¼Œçœ‹è¾“å‡ºç»“æœ
c(1,"a")
c(TRUE,"a")
c(1,TRUE)

# 4.ç”¨å‡½æ•°è®¡ç®—å‘é‡gçš„é•¿åº¦
load("gands.Rdata")
length(g)
# 5.ç­›é€‰å‡ºå‘é‡gä¸­ä¸‹æ ‡ä¸ºå¶æ•°çš„åŸºå› åã€‚
seq(2,100,2)
g[seq(2,100,2)]
# 6.å‘é‡gä¸­æœ‰å¤šå°‘ä¸ªå…ƒç´ åœ¨å‘é‡sä¸­å­˜åœ¨(è¦æ±‚ç”¨å‡½æ•°è®¡ç®—å‡ºå…·ä½“ä¸ªæ•°)ï¼Ÿå°†è¿™äº›å…ƒç´ ç­›é€‰å‡ºæ¥
# æç¤ºï¼š%in%
table(g %in% s)
g[g %in% s]
# 7.ç”Ÿæˆ10ä¸ªéšæœºæ•°: rnorm(n=10,mean=0,sd=18)ï¼Œç”¨å‘é‡å–å­é›†çš„æ–¹æ³•ï¼Œå–å‡ºå…¶ä¸­å°äº-2çš„å€¼
z = rnorm(n=10,mean=0,sd=18)
z
z[z<-2]
z

z = rnorm(n=10,mean=0,sd=18)
z
z[z< -2]
z[z<(-2)]

```

### 2.æ•°æ®æ¡†ï¼ŒçŸ©é˜µï¼Œåˆ—è¡¨

```{r}
#é‡ç‚¹ï¼šæ•°æ®æ¡†
#1.æ•°æ®æ¡†æ¥æº
# ï¼ˆ1ï¼‰ç”¨ä»£ç æ–°å»º
# ï¼ˆ2ï¼‰ç”±å·²æœ‰æ•°æ®è½¬æ¢æˆ–å¤„ç†å¾—åˆ°
# ï¼ˆ3ï¼‰è¯»å–è¡¨æ ¼æ–‡ä»¶
# ï¼ˆ4ï¼‰Rè¯­è¨€å†…ç½®æ•°æ®

#2.æ–°å»ºå’Œè¯»å–æ•°æ®æ¡†
df1 <- data.frame(gene   = paste0("gene",1:4),
                 change  = rep(c("up","down"),each = 2),
                 score   = c(5,3,-2,-4))
df1

df2 <- read.csv("gene.csv")
df2

#3.æ•°æ®æ¡†å±æ€§
#
dim(df1)
nrow(df1)
ncol(df1)
#
rownames(df1)
colnames(df1)

#4.æ•°æ®æ¡†å–å­é›†
df1$gene  #åˆ æ‰scoreï¼ŒæŒ‰tabé”®è¯•è¯•
mean(df1$score)

## æŒ‰åæ ‡
df1[2,2]
df1[2,]
df1[,2]
df1[c(1,3),1:2]
## æŒ‰åå­—
df1[,"gene"]
df1[,c('gene','change')]
## æŒ‰æ¡ä»¶ï¼ˆé€»è¾‘å€¼ï¼‰
df1[df1$score>0,]

## ä»£ç æ€ç»´
#å¦‚ä½•å–æ•°æ®æ¡†çš„æœ€åä¸€åˆ—ï¼Ÿ
df1[,3]
df1[,ncol(df1)]
#å¦‚ä½•å–æ•°æ®æ¡†é™¤äº†æœ€åä¸€åˆ—ä»¥å¤–çš„å…¶ä»–åˆ—ï¼Ÿ
df1[,-ncol(df1)]

#ç­›é€‰score > 0çš„åŸºå› 
df1[df1$score > 0,1]
df1$gene[df1$score > 0]

#5.æ•°æ®æ¡†ä¿®æ”¹

#æ”¹ä¸€ä¸ªæ ¼
df1[3,3] <- 5
df1
#æ”¹ä¸€æ•´åˆ—
df1$score <- c(12,23,50,2)     
df1
#ï¼Ÿ
df1$p.value <- c(0.01,0.02,0.07,0.05) 
df1

#æ”¹è¡Œåå’Œåˆ—å
rownames(df1) <- c("r1","r2","r3","r4")
#åªä¿®æ”¹æŸä¸€è¡Œ/åˆ—çš„å
colnames(df1)[2] <- "CHANGE"

#6.ä¸¤ä¸ªæ•°æ®æ¡†çš„è¿æ¥
test1 <- data.frame(name = c('jimmy','nicker','Damon','Sophie'), 
                    blood_type = c("A","B","O","AB"))
test1
test2 <- data.frame(name = c('Damon','jimmy','nicker','tony'),
                    group = c("group1","group1","group2","group2"),
                    vision = c(4.2,4.3,4.9,4.5))
test2

test3 <- data.frame(NAME = c('Damon','jimmy','nicker','tony'),
                    weight = c(140,145,110,138))
test3
merge(test1,test2,by="name") #å–äº¤é›†äº†
merge(test1,test3,by.x = "name",by.y = "NAME")

##### çŸ©é˜µå’Œåˆ—è¡¨
m <- matrix(1:9, nrow = 3)
colnames(m) <- c("a","b","c") #åŠ åˆ—å
m
m[2,]
m[,1]
m[2,3]
m[2:3,1:2]
m
t(m)
as.data.frame(m)
#åˆ—è¡¨
l <- list(m1 = matrix(1:9, nrow = 3),
          m2 = matrix(2:9, nrow = 2))
l

l[[2]]
l$m1

# è¡¥å……ï¼šå…ƒç´ çš„åå­—

scores = c(100,59,73,95,45)
names(scores) = c("jimmy","nicker","Damon","Sophie","tony")
scores
scores["jimmy"]
scores[c("jimmy","nicker")]

names(scores)[scores>60]

# åˆ é™¤ 
rm(l)
rm(df1,df2)
rm(list = ls()) 

# matchç»ƒä¹ é¢˜
load("matchtest.Rdata")


```

### 3.å¾ªç¯å’Œå†™å‡½æ•°

```{r}
jimmy <- function(a,b,m = 2){
  (a+b)^m
}
jimmy(a = 1,b = 2)
jimmy(1,2)
jimmy(3,6)
jimmy(3,6,-2)

#å¤ä¹ ï¼šç»˜å›¾å‡½æ•°plot()
par(mfrow = c(2,2)) #æŠŠç”»æ¿åˆ†æˆå››å—ï¼Œä¸¤è¡Œä¸¤åˆ—
#å¦‚æœæŠ¥é”™ï¼ŒæŠŠå³ä¸‹è§’ç”»æ¿æ‹‰å¤§ä¸€ç‚¹å³å¯
x = c(2,5,6,2,9);plot(x)
x = seq(2,80,4);plot(x)
x = rnorm(10);plot(x)
x = iris$Sepal.Length;plot(x)

#æ€è€ƒï¼šplotç”»irisçš„å‰å››åˆ—ï¼Ÿ
plot(iris[,1],col = iris[,5])
plot(iris[,2],col = iris[,5])
plot(iris[,3],col = iris[,5])
plot(iris[,4],col = iris[,5])

#å½“ä¸€ä¸ªä»£ç éœ€è¦å¤åˆ¶ç²˜è´´ä¸‰æ¬¡ï¼Œå°±åº”è¯¥å†™æˆå‡½æ•°æˆ–ä½¿ç”¨å¾ªç¯

jimmy <- function(i){
  plot(iris[,i],col=iris[,5])
}

jimmy(1)
jimmy(2)
jimmy(3)
jimmy(4)
```

### 4.RåŒ…å®‰è£…åŠä½¿ç”¨

```{r}
# RåŒ…å®‰è£…

options("repos"=c(CRAN="http://mirrors.tuna.tsinghua.edu.cn/CRAN/"))
options(BioC_mirror="http://mirrors.ustc.edu.cn/bioc/")
# æ¸…åé•œåƒ
# http://mirrors.tuna.tsinghua.edu.cn/CRAN/
# http://mirrors.tuna.tsinghua.edu.cn/bioconductor/
  
# ä¸­ç§‘å¤§é•œåƒ
# http://mirrors.ustc.edu.cn/CRAN/
# http://mirrors.ustc.edu.cn/bioc/

install.packages("tidyr")
install.packages('BiocManager')
#BiocManager::installå¯ä»¥å®‰è£…ä¸‰ä¸ªæ¥æºçš„åŒ…
BiocManager::install("ggplot2")
BiocManager::install("jmzeng1314/idmap1")
BiocManager::install("tidyr")
install.packages('devtools')
devtools::install_github("jmzeng1314/idmap1") #æ‹¬å·é‡Œå†™ä½œè€…ç”¨æˆ·ååŠ åŒ…å



  
library(tidyr)
require(tidyr)

# åˆ†æƒ…å†µè®¨è®º

if(!require(stringr))install.packages("stringr")

# è·å–å¸®åŠ©
?seq
library(stringr)
browseVignettes("stringr")
ls("package:stringr")


```

### 5.æ–‡ä»¶è¯»å†™

```{r}
#æ–‡ä»¶è¯»å†™éƒ¨åˆ†
#1.è¯»å–ex1.txt
ex1 <- read.table("ex1.txt")
ex1 <- read.table("ex1.txt",header = T)
#2.è¯»å–ex2.csv
ex2 <- read.csv("ex2.csv")
ex2 <- read.csv("ex2.csv",row.names = 1,check.names = F)

#æ³¨æ„ï¼šæ•°æ®æ¡†ä¸å…è®¸é‡å¤çš„è¡Œå
rod = read.csv("rod.csv",row.names = 1)
rod = read.csv("rod.csv")

#3.è¯»å–soft.txt
soft <- read.table("soft.txt")
soft <- read.table("soft.txt",header = T,fill = T) #å…¶å®ä¸å¯¹
soft2 <- read.table("soft.txt",header = T,sep = "\t")

#4.soft çš„è¡Œæ•°åˆ—æ•°æ˜¯å¤šå°‘ï¼Ÿåˆ—åæ˜¯ä»€ä¹ˆ
dim(soft)
colnames(soft)
#5.å°†softå¯¼å‡ºä¸ºcsv
write.csv(soft,file = "soft.csv")
#6.å°†softä¿å­˜ä¸ºRdataå¹¶åŠ è½½ã€‚
save(soft,file = "soft.Rdata")
rm(list = ls())
load(file = "soft.Rdata")


```

### 6.åŸºç¡€ä½œå›¾

#### 1.ä½œå›¾åˆ†ä¸‰ç±»

```{r}


#1.åŸºç¡€åŒ… ç•¥æ˜¾é™ˆæ—§ äº†è§£ä¸€ä¸‹
plot(iris[,1],iris[,3],col = iris[,5]) 
text(6.5,4, labels = 'hello')

dev.off() #å…³é—­ç”»æ¿

#2.ggplot2 ä¸­åšåŠ›é‡ï¼Œè¯­æ³•æœ‰ä¸ªæ€§
library(ggplot2)
ggplot(data = iris)+
  geom_point(mapping = aes(x = Sepal.Length,
                           y = Petal.Length,
                           color = Species))

#3.ggpubr æ–°æ‰‹å‹å¥½å‹ ggplot2ç®€åŒ–å’Œç¾åŒ– è¤’è´¬ä¸ä¸€
library(ggpubr)
ggscatter(iris,
          x="Sepal.Length",
          y="Petal.Length",
          color="Species")

```

#### 2.ggplot2

```{r}
library(ggplot2)
#1.å…¥é—¨çº§ç»˜å›¾æ¨¡æ¿ï¼šä½œå›¾æ•°æ®ï¼Œæ¨ªçºµåæ ‡
ggplot(data = iris)+
  geom_point(mapping = aes(x = Sepal.Length,
                           y = Petal.Length))
#2.å±æ€§è®¾ç½®ï¼ˆé¢œè‰²ã€å¤§å°ã€é€æ˜åº¦ã€ç‚¹çš„å½¢çŠ¶ï¼Œçº¿å‹ç­‰ï¼‰

#2.1 æ‰‹åŠ¨è®¾ç½®ï¼Œéœ€è¦è®¾ç½®ä¸ºæœ‰æ„ä¹‰çš„å€¼

ggplot(data = iris) + 
  geom_point(mapping = aes(x = Sepal.Length,
                           y = Petal.Length), 
             color = "blue")

ggplot(data = iris) + 
  geom_point(mapping = aes(x = Sepal.Length, y = Petal.Length), 
             size = 5,     # ç‚¹çš„å¤§å°5mm
             alpha = 0.5,  # é€æ˜åº¦ 50%
             shape = 21)  # ç‚¹çš„å½¢çŠ¶

#2.2 æ˜ å°„ï¼šæŒ‰ç…§æ•°æ®æ¡†çš„æŸä¸€åˆ—æ¥å®šä¹‰å›¾çš„æŸä¸ªå±æ€§
ggplot(data = iris)+
  geom_point(mapping = aes(x = Sepal.Length,
                           y = Petal.Length,
                           color = Species))

## Q1 èƒ½ä¸èƒ½è‡ªè¡ŒæŒ‡å®šæ˜ å°„çš„å…·ä½“é¢œè‰²ï¼Ÿ

ggplot(data = iris)+
  geom_point(mapping = aes(x = Sepal.Length,
                           y = Petal.Length,
                           color = Species))+
  scale_color_manual(values = c("blue","grey","red"))

## Q2 åŒºåˆ†colorå’Œfillä¸¤ä¸ªå±æ€§
### Q2-1 ç©ºå¿ƒå½¢çŠ¶å’Œå®å¿ƒå½¢çŠ¶éƒ½ç”¨colorè®¾ç½®é¢œè‰²
ggplot(data = iris)+
  geom_point(mapping = aes(x = Sepal.Length,
                           y = Petal.Length,
                           color = Species),
             shape = 17) #17å·ï¼Œå®å¿ƒçš„ä¾‹å­

ggplot(data = iris)+
  geom_point(mapping = aes(x = Sepal.Length,
                           y = Petal.Length,
                           color = Species),
             shape = 2) #2å·ï¼Œç©ºå¿ƒçš„ä¾‹å­
### Q2-2 æ—¢æœ‰è¾¹æ¡†åˆæœ‰å†…å¿ƒçš„ï¼Œæ‰éœ€è¦colorå’Œfillä¸¤ä¸ªå‚æ•°

ggplot(data = iris)+
  geom_point(mapping = aes(x = Sepal.Length,
                           y = Petal.Length,
                           color = Species),
             shape = 24,
             fill = "black") #24å·ï¼ŒåŒè‰²çš„ä¾‹å­

#3.åˆ†é¢
ggplot(data = iris) + 
  geom_point(mapping = aes(x = Sepal.Length, y = Petal.Length)) + 
  facet_wrap(~ Species) 
#åŒåˆ†é¢
dat = iris
dat$Group = sample(letters[1:5],150,replace = T)
ggplot(data = dat) + 
  geom_point(mapping = aes(x = Sepal.Length, y = Petal.Length)) + 
  facet_grid(Group ~ Species) 

#4.å‡ ä½•å¯¹è±¡

#å±€éƒ¨è®¾ç½®å’Œå…¨å±€è®¾ç½®

ggplot(data = iris) + 
  geom_smooth(mapping = aes(x = Sepal.Length, 
                          y = Petal.Length))+
  geom_point(mapping = aes(x = Sepal.Length, 
                           y = Petal.Length))

ggplot(data = iris,mapping = aes(x = Sepal.Length, y = Petal.Length))+
  geom_smooth()+
  geom_point()

#5.ç»Ÿè®¡å˜æ¢-ç›´æ–¹å›¾
View(diamonds)
table(diamonds$cut)

ggplot(data = diamonds) + 
  geom_bar(mapping = aes(x = cut))

ggplot(data = diamonds) + 
  stat_count(mapping = aes(x = cut))

#ç»Ÿè®¡å˜æ¢ä½¿ç”¨åœºæ™¯
#5.1.ä¸ç»Ÿè®¡ï¼Œæ•°æ®ç›´æ¥åšå›¾
fre = as.data.frame(table(diamonds$cut))
fre

ggplot(data = fre) +
  geom_bar(mapping = aes(x = Var1, y = Freq), stat = "identity")
#5.2countæ”¹ä¸ºprop
ggplot(data = diamonds) + 
  geom_bar(mapping = aes(x = cut, y = ..prop.., group = 1))


#6.ä½ç½®å…³ç³»

# 6.1æŠ–åŠ¨çš„ç‚¹å›¾
ggplot(data = iris,mapping = aes(x = Species, 
                                 y = Sepal.Width,
                                 fill = Species)) + 
  geom_boxplot()+
  geom_point()

ggplot(data = iris,mapping = aes(x = Species, 
                                 y = Sepal.Width,
                                 fill = Species)) + 
  geom_boxplot()+
  geom_jitter()

# 6.2å †å ç›´æ–¹å›¾
ggplot(data = diamonds) + 
  geom_bar(mapping = aes(x = cut,fill=clarity))

# 6.3 å¹¶åˆ—ç›´æ–¹å›¾
ggplot(data = diamonds) + 
  geom_bar(mapping = aes(x = cut, fill = clarity), position = "dodge")

#7.åæ ‡ç³»

#ç¿»è½¬coord_flip()

ggplot(data = mpg, mapping = aes(x = class, y = hwy)) + 
  geom_boxplot() +
  coord_flip()
#æåæ ‡ç³»coord_polar()
bar <- ggplot(data = diamonds) + 
  geom_bar(
    mapping = aes(x = cut, fill = cut), 
    width = 1
  ) + 
  theme(aspect.ratio = 1) +
  labs(x = NULL, y = NULL)
bar
bar + coord_flip()
bar + coord_polar()



```

#### 3.ggpubr

```{r}
# ggpubr æœä»£ç ç›´æ¥ç”¨ï¼ŒåŸºæœ¬ä¸éœ€è¦ç³»ç»Ÿå­¦ä¹ 

# sthdaä¸Šæœ‰å¤§é‡ggpubrå‡ºçš„å›¾
library(ggpubr)
ggscatter(iris,x="Sepal.Length",
          y="Petal.Length",
          color="Species")

p <- ggboxplot(iris, x = "Species", 
               y = "Sepal.Length",
               color = "Species", 
               shape = "Species",
               add = "jitter")
p
my_comparisons <- list( c("setosa", "versicolor"), 
                        c("setosa", "virginica"), 
                        c("versicolor", "virginica") )
p + stat_compare_means(comparisons = my_comparisons)+ # Add pairwise comparisons p-value
  stat_compare_means(label.y = 9) 

```

#### 4.å›¾ç‰‡ä¿å­˜çš„ä¸‰ç§æ–¹æ³•

```{r}
#å›¾ç‰‡ä¿å­˜çš„ä¸‰ç§æ–¹æ³•

#1.åŸºç¡€åŒ…ä½œå›¾çš„ä¿å­˜
pdf("iris_box_ggpubr.pdf")
boxplot(iris[,1]~iris[,5])
text(6.5,4, labels = 'hello')
dev.off()

#2.ggplotç³»åˆ—å›¾(åŒ…æ‹¬ggpubr)é€šç”¨çš„ç®€ä¾¿ä¿å­˜ ggsave
p <- ggboxplot(iris, x = "Species", 
               y = "Sepal.Length",
               color = "Species", 
               shape = "Species",
               add = "jitter")
ggsave(p,filename = "iris_box_ggpubr.png")

#3.eofficeåŒ… å¯¼å‡ºä¸ºppt,å…¨éƒ¨å…ƒç´ éƒ½æ˜¯å¯ç¼–è¾‘æ¨¡å¼
library(eoffice)
topptx(p,"iris_box_ggpubr.pptx")

#https://mp.weixin.qq.com/s/p7LLLvzR5LPgHhuRGhYQBQ

```

### 7.tidyverseåŸºç¡€

#### 1.tidyverseè½¯ä»¶å®‰è£…

```{r}
options("repos" = c(CRAN="http://mirrors.tuna.tsinghua.edu.cn/CRAN/"))
if(!require(tidyr))install.packages("tidyr",update = F,ask = F)
if(!require(dplyr))install.packages("dplyr",update = F,ask = F)
if(!require(stringr))install.packages('stringr',update = F,ask = F)
if(!require(tibble))install.packages('tibble',update = F,ask = F)
library(tidyr)
library(dplyr)
library(stringr)
library(tibble)
#æˆ–è€…
if(!require(tidyr))install.packages("tidyverse",update = F,ask = F)
library(tidyverse)

```

#### 2.ç©è½¬å­—ç¬¦ä¸²stringr

```{r}
rm(list = ls())
if(!require(stringr))install.packages('stringr')
library(stringr)

x <- "The birch canoe slid on the smooth planks."
x
### 1.æ£€æµ‹å­—ç¬¦ä¸²é•¿åº¦
str_length(x)
length(x)
### 2.å­—ç¬¦ä¸²æ‹†åˆ†
str_split(x," ")
x2 = str_split(x," ")[[1]];x2

y = c("jimmy 150","nicker 140","tony 152")
str_split(y," ")
str_split(y," ",simplify = T)

### 3.æŒ‰ä½ç½®æå–å­—ç¬¦ä¸²
str_sub(x,5,9)

### 4.å­—ç¬¦æ£€æµ‹
str_detect(x2,"h")
str_starts(x2,"T")
str_ends(x2,"e")
### 5.å­—ç¬¦ä¸²æ›¿æ¢
x2
str_replace(x2,"o","A")
str_replace_all(x2,"o","A")

### 6.å­—ç¬¦åˆ é™¤
x
str_remove(x," ")
str_remove_all(x," ")

```

#### 3.ç©è½¬æ•°æ®æ¡†dplyr

```{r}
test <- iris[c(1:2,51:52,101:102),]
rownames(test) =NULL # å»æ‰è¡Œåï¼ŒNULLæ˜¯â€œä»€ä¹ˆéƒ½æ²¡æœ‰â€
test

# arrangeï¼Œæ•°æ®æ¡†æŒ‰ç…§æŸä¸€åˆ—æ’åº

library(dplyr)
arrange(test, Sepal.Length) #ä»å°åˆ°å¤§
arrange(test, desc(Sepal.Length)) #ä»å¤§åˆ°å°

# distinctï¼Œæ•°æ®æ¡†æŒ‰ç…§æŸä¸€åˆ—å»é‡å¤
distinct(test,Species,.keep_all = T)

# mutateï¼Œæ•°æ®æ¡†æ–°å¢ä¸€åˆ—,æˆ–è€…ä¿®æ”¹åŸæ¥åœ°åˆ—
mutate(test, new = Sepal.Length * Sepal.Width)

# è¿ç»­çš„æ­¥éª¤

# 1.å¤šæ¬¡èµ‹å€¼ï¼Œäº§ç”Ÿå¤šä¸ªå˜é‡

x1 = filter(iris,Sepal.Width>3)
x2 = select(x1, Sepal.Length,Sepal.Width)
x3 = arrange(x2,Sepal.Length)

# 2.ç®¡é“ç¬¦å·ä¼ é€’ï¼Œç®€æ´æ˜äº†
x = iris %>% 
  filter(Sepal.Width>3) %>% 
  select(Sepal.Length,Sepal.Width)%>%
  arrange(Sepal.Length)

# 3. åµŒå¥—ï¼Œä»£ç ä¸æ˜“è¯»
arrange(select(filter(iris,Sepal.Width>3),
               Sepal.Length,Sepal.Width),
        Sepal.Length)

```

## 




# å°æŠ€å·§

### ä¸€äº›æŠ¥é”™çš„råŒ…ï¼Œå¯ä»¥ä½¿ç”¨condaå®‰è£…

```{bash}
#ä¾‹å¦‚ï¼Œä½†ä¼šæ¯”è¾ƒæ…¢
conda install -c bioconda bioconductor-clusterprofiler
conda install bioconda::bioconductor-clusterprofiler
conda install -c conda-forge r-tidyverse
```

### è°ƒç”¨å…¶ä»–condaç¯å¢ƒä¸‹çš„RåŒ…

```{r}
#æ–¹æ³•ä¸€ï¼šæŒ‡å®šRåŒ…ä½ç½®å»è·¨ç¯å¢ƒè¯»å–è¯»å–
#å¦‚æœä¸€å€‹ç¯å¢ƒä¸‹å®‰è£…æ²¡æˆåŠŸï¼Œå¦ä¸€ä¸ªç¯å¢ƒæˆåŠŸäº†ï¼Œå¯ä»¥ç›´æ¥æŒ‡å®šRåŒ…ä½ç½®å»è·¨ç¯å¢ƒè¯»å–è¯»å–
library("tidyverse", lib.loc="/home/hekun/anaconda3/envs/RStudio/lib/R/library/")
library("clusterProfiler", lib.loc="/home/hekun/anaconda3/envs/RStudio/lib/R/library/")
#æ–¹æ³•äºŒï¼šä¸´æ—¶æ·»åŠ .libPaths()çš„æ–°çš„è¯»å–è·¯å¾„ï¼Œé‡å¯Råå¤±æ•ˆ
.libPaths()
.libPaths("/home/hekun/anaconda3/envs/RStudio/lib/R/library/")

#æ–¹æ³•ä¸‰ï¼šæ°¸ä¹…æ·»åŠ .libPaths()çš„æ–°çš„è¯»å–è·¯å¾„ï¼Œæ”¹.Rprofileï¼Œåœ¨.Rprofileä¸­æ·»åŠ ä»¥ä¸‹ä»£ç è¡Œæ¥è®¾ç½®é¢å¤–çš„åº“è·¯å¾„ï¼š
vim /home/hekun/.Rprofile
.libPaths(c("/home/hekun/anaconda3/envs/RStudio/lib/R/library", .libPaths()))
```

### å®‰è£…æŒ‡å®šç‰ˆæœ¬åŒ…

```{r}
#å®‰è£…ç”¨æ¥è£…æ—§åŒ…çš„åŒ…
install.packages("remotes")
#å®‰è£…è¦è£…çš„å·²å®‰è£…çš„åŒ…
remove.packages("matrixStats")
remotes::install_version("matrixStats", version = "1.1.0")


```

### Rç»˜å›¾æŒ‡ä»¤

æ™®é€šå›¾ç»˜åˆ¶
```{r}
# è®¾ç½®ä¿å­˜è·¯å¾„ä¸ºPDFæ–‡ä»¶
pdf("samples_cor.pdf", width = 10, height = 10) 

# ç»˜åˆ¶çƒ­å›¾
pheatmap(sample_cor, 
         cluster_rows = F, cluster_cols = F, # ä¸èšç±»
         cellwidth = 15, cellheight = 15, # cell å¤§å°
         border_color = "white", # è¾¹æ¡†é¢œè‰²
         fontsize = 8, # å­—ä½“å¤§å°
         angle_col = 45, # åˆ—å€¾æ–œ
         display_numbers = T, # æ˜¾ç¤ºæ•°å€¼
         fontsize_number = 5) # æ•°å€¼å­—ä½“å¤§å°

# å…³é—­è®¾å¤‡ï¼Œç¡®ä¿æ–‡ä»¶ä¿å­˜
dev.off()


```

